File: go_files/ex11.go
package main

import "fmt"

// Definindo a interface Funcionario
type Funcionario interface {
	CalcularSalario() float64
}

type FuncionarioHorista struct {
	nome string
	horasTrabalhadas float64
	valorHora float64
}

func (f FuncionarioHorista) CalcularSalario() float64 {
	return f.horasTrabalhadas * f.valorHora
}

type FuncionarioAssalariado struct {
	nome        string
	salarioFixo float64
}

func (f FuncionarioAssalariado) CalcularSalario() float64 {
	return f.salarioFixo
}

func imprimirSalario(funcionario Funcionario, nome string) {
	fmt.Printf("Salário de %s: R$ %.2f\n", nome, funcionario.CalcularSalario())
}

func main() {
	funcHorista := FuncionarioHorista{nome: "João", horasTrabalhadas: 40, valorHora: 20}
	funcAssalariado := FuncionarioAssalariado{nome: "Maria", salarioFixo: 3000}

	imprimirSalario(funcHorista, "João")
	imprimirSalario(funcAssalariado, "Maria")
}

File: python_files/ex11.py
from abc import ABC, abstractmethod

class Funcionario(ABC):
    def __init__(self, nome):
        self.nome = nome
    
    @abstractmethod
    def calcular_salario():
        pass


class FuncionarioHorista(Funcionario):
    def __init__(self, nome, horas_trabalhadas, valor_hora):
        super().__init__(nome)
        self.horas_trabalhadas = horas_trabalhadas
        self.valor_hora = valor_hora

    def calcular_salario(self):
        return self.horas_trabalhadas * self.valor_hora
    

class FuncionarioAssalariado(Funcionario):
    def __init__(self, nome, salario_fixo):
        super().__init__(nome)
        self.salario_fixo = salario_fixo

    def calcular_salario(self):
        return self.salario_fixo
    
func_horista = FuncionarioHorista("João", 40, 20)
func_assalariado = FuncionarioAssalariado("Maria", 3000)

print(f"Salário de {func_horista.nome}: R$ {func_horista.calcular_salario()}")
print(f"Salário de {func_assalariado.nome}: R$ {func_assalariado.calcular_salario()}")